features/step_definitions/web_steps.rb

```diff
@@ -30,8 +30,12 @@ def with_scope(locator)
   end
 end
 
-When /^(?:|I |they |the buyer )follow( invisible)? "([^"]*)"(?: within "([^"]*)")?$/ do |invisible, link, selector|
+When /^(?:|I |they |the buyer )follow( any)?( invisible)? "([^"]*)"(?: within "([^"]*)")?$/ do |any, invisible, link, selector|
   with_scope(selector) do
+    # there must be a capybara bug because assert_link fails with
+    # Unused parameters passed to Capybara::Queries::SelectorQuery : [:link, "..."]
+    # assert_link(link, exact: true, visible: !invisible, count: 1) unless any
+    assert_selector(:link, link, exact: true, visible: !invisible, count: 1) unless any
```

Comment:
assert_link is buggy, have you tried with `assert has_link?(...)`?

I really don't like this idea of the "any" and I don't like this step at all. What sense does it make to assert the selector if then it's going to be clicked? `click_link` will assert the selector already 🤷🏻‍♂️.

Reply:
> What sense does it make to assert the selector if then it's going to be clicked? 

It doesn't merely assert that the link exists, but that there is only one link, not multiple matching the criteria.

The `any` variant is when we know that there are multiple matching links but we don't care which one would be clicked.

In this way, we will be warned when there are multiple matching links while writing the test and consider whether we care which one is clicked or not. This will also make tests less dependent on page sort order.

> Moreover this step is a duplication of They follow in combination with {} within {selector}.

This is not a copy, it is changing that same step.

Reply:
Why not using `has_link?`?

Reply:
IMHO trying to click "any" link smells bad. It smells like we (tester) are not sure what there is to be tested. We should be as much specific as we can. If there are "many" equal links, then it should not matter at all and the step should still be "I follow X" and click the first one I find. If there 2 link that "look like" the same but they are not, then we should be more specific "I follow the link to do X" or "I follow the link within the nav", etc.

Reply:
>Why not using has_link??

`assert_selector` already provides a proper error message. I will need to emulate this with `has_link?` and it will be uglier and more complicated.

> IMHO trying to click "any" link smells bad.

In most cases - I agree. When we have multiple equal elements like multiple custom IDs, then it is fine. Why should I care which one was clicked? But you may suggest specific improvements for the scenarios where `any` was used.

Note that this is no change in behavior, as previously this is what happened. Now it is made obvious. That is the major change. And it is good to be obvious. Because if we don't like, we may fix. Previously nobody knew.

Reply:
I agree with @josemigallas here... I don't like this "any" either...

If anything, I would replace it with "first", because I think that would reflect the reality better. I believe capybara will not randomly select the link, but just stop at the first match?

Reply:
OK, it seems that I am late to the party :sweat_smile:

Reply:
I don't know if first is guaranteed. Maybe it is documented. But relying on order of elements is even worse to me than stating `any`. `any` might be the first one as well. But `first` is not `any`.

If we rely on order, then UI changes may require refactoring a lot of tests. So `any` is kind of a warning that it could potentially be any link matching that selector. And it is good nobody likes it, because it shouldn't be used in most if not all cases :)

As I said above, we have always had this, nobody noticed it until now. And we probably have the same with button clicking and other places.